apiserver/debuglog.go

```diff
@@ -182,29 +182,37 @@ func (s *debugLogSocketImpl) sendLogRecord(record *params.LogMessage) error {
 
 // debugLogParams contains the parsed debuglog API request parameters.
 type debugLogParams struct {
-	startTime     time.Time
-	maxLines      uint
-	fromTheStart  bool
-	noTail        bool
-	backlog       uint
-	filterLevel   loggo.Level
-	includeEntity []string
-	excludeEntity []string
-	includeModule []string
-	excludeModule []string
-	includeLabel  []string
-	excludeLabel  []string
+	startTime      time.Time
+	fromTheStart   bool
+	noTail         bool
+	latestLogCount uint
+	filterLevel    loggo.Level
+	includeEntity  []string
+	excludeEntity  []string
+	includeModule  []string
+	excludeModule  []string
+	includeLabel   []string
+	excludeLabel   []string
 }
 
 func readDebugLogParams(queryMap url.Values) (debugLogParams, error) {
 	var params debugLogParams
 
+	if value := queryMap.Get("backlog"); value != "" {
+		num, err := strconv.ParseUint(value, 10, 64)
+		if err != nil {
+			return params, errors.Errorf("backlog value %q is not a valid unsigned number", value)
+		}
+		params.latestLogCount = uint(num)
+	}
+
 	if value := queryMap.Get("maxLines"); value != "" {
 		num, err := strconv.ParseUint(value, 10, 64)
 		if err != nil {
 			return params, errors.Errorf("maxLines value %q is not a valid unsigned number", value)
 		}
-		params.maxLines = uint(num)
+		params.noTail = true
```

Comment:
Hm are you referring to the statement "params.noTail = true" ? I drafted this solution based on our 1-1 discussion previously to not include the additional field "maxLimit" in the state and to mimick "limit" behaviour as --lines --notail. Do you mind sharing how this breaks the expectations of clients? Perhaps then I could try to see if there's another workaround to achieve both the constraints

Reply:
This server side code accepts `params.noTail` as input according to what the client has specified. Given the current semantics and expectations, this server side code should not attempt to set input parameters - it should validate what is gets passed is semantically valid. If someone does `--limit` then that means we should ask for `maxLines` and we should also expect  `noTail` is true when making the api call. Otherwise, what happens if someone tries `--limit --tail` - if that's a problem, which it appears to be because the two parameters conflict, we should error, both in the debug-log command and also here server side as a backstop. Ultimately if we rework all this in 4.0 we would have composable primitives - maxLines, tail yes/no, replay yes/no - and there'd be no conflicts.

Reply:
Do we have enough checks in debuglog.go to ensure we error when the user attempts to supply conflicting parameters?

Reply:
thanks Ian for the feedback, from what I understood the main problem seems to be if user does:

debug-log --limit --lines
debug-log --limit --tail

but the current implementation seems to already allows for conflicting flags when we allow users to type --tail and --notail, however our readDebugLogParams doesn't seem to check for --tail. I think we could throw an error if user does types  --limit --lines for now? I'm not sure how we could completely handle user's expectations given that we allow --tail and --notail, perhaps we could discuss on this if needed.